python/pyspark/mllib/linalg.py

```diff
@@ -1152,9 +1156,385 @@ def sparse(numRows, numCols, colPtrs, rowIndices, values):
         return SparseMatrix(numRows, numCols, colPtrs, rowIndices, values)
 
 
+class DistributedMatrix(object):
+    """Represents a distributively stored matrix backed by one or more RDDs."""
+    def numRows(self):
+        """Get or compute the number of rows."""
+        raise NotImplementedError
+
+    def numCols(self):
+        """Get or compute the number of cols."""
+        raise NotImplementedError
+
+
+class DistributedMatrices(object):
+    """Factory methods for distributed matrices."""
+    @staticmethod
+    def rowMatrix(rows, numRows=0, numCols=0):
+        """
+        Create a RowMatrix.
+
+        :param rows: An RDD of Vectors.
+        """
+        javaRowMatrix = callMLlibFunc("createRowMatrix", rows, long(numRows), int(numCols))
+        jrm = JavaModelWrapper(javaRowMatrix)
+        return RowMatrix(jrm)
+
+    @staticmethod
+    def indexedRowMatrix(rows, numRows=0, numCols=0):
+        """
+        Create an IndexedRowMatrix.
+
+        :param rows: An RDD of IndexedRows or (long, Vector) tuples.
+        """
+        # We use DataFrames for serialization of IndexedRows from Python, so convert the RDD to a
+        # DataFrame. This will convert each IndexedRow to a Row containing the 'index' and 'vector'
+        # values, which can both be easily serialized.  We will convert back to IndexedRows on the
+        # Scala side.
+        javaIndexedRowMatrix = callMLlibFunc("createIndexedRowMatrix", rows.toDF(),
```

Comment:
The idea is that the `IndexedRow` Python type isn't directly serializable to a format that can be read by the Java side, and vice-versa.  Any of these custom data structures require explicit serialization, which can be found in the `SerDe` Scala class at the bottom of `PythonMLlibAPI.scala`, as well as a bit of representation logic on the Python side.  Prior to `DataFrame`s, this was done for any structure that needed to be passed between the languages, so we already have explicit conversions for `DenseVector`, `DenseMatrix`, `LabeledPoint`, `Rating`, etc.  The `DataFrame` structure lets us pull the contents out of these custom data structures and store them in a a generic `Row`.  The contents are often simpler types, and `DataFrame`s can serialization "primitives" such as `Int`, `Long`, `Boolean`, etc., as well as a few custom types such as `Vector` and `Matrix` types.  So the preferred way of serialization moving forward is to use `DataFrame`s for serialization when possible.  For the `IndexedRow` type, it just contains a `long` and a `Vector`, both of which can be easily serialized with `DataFrame`s, even though the `IndexedRow` itself isn't cross-language serializable.  Converting the `RDD` of `IndexedRow`s in Python to a `DataFrame` creates a `DataFrame` with `Row`s containing just the `long` and `Vector` from the `IndexedRow`.